toonz/sources/toonz/xsheetdragtool.cpp

```diff
@@ -1595,22 +1597,53 @@ class ColumnMoveDragTool final : public XsheetGUI::DragTool {
     if (indices.empty()) return;
     m_firstCol = m_lastCol = *indices.begin();
     assert(m_firstCol >= 0);
-    m_offset = m_firstCol - col;
+    m_origOffset = m_offset = m_firstCol - col;
     assert(m_lastCol == *indices.begin());
     getViewer()->update();
   }
   void onDrag(const CellPosition &pos) override {
     int col                     = pos.layer();
     TColumnSelection *selection = getViewer()->getColumnSelection();
-    std::set<int> indices       = selection->getIndices();
+    TApp *app                   = TApp::instance();
+    TXsheet *xsh                = app->getCurrentXsheet()->getXsheet();
+
+    std::set<int> indices = selection->getIndices();
     if (indices.empty()) return;
 
     assert(m_lastCol == *indices.begin());
 
+    int origCol      = col;
     if (col < 0) col = 0;
-    if (col == m_lastCol) return;
-    int dCol  = col - m_lastCol;
-    m_lastCol = col;
+    int dCol         = col - (m_lastCol - m_offset);
+    if (origCol < 0) dCol += origCol;
+
+    std::set<int> ii;
+    int newBegin = *indices.begin() + dCol;
+    if (newBegin < 0) {
+      newBegin *= -1;
+      for (int x = 0; x < newBegin; x++) ii.insert(x);
+      ColumnCmd::insertEmptyColumns(ii);
+      selection->selectNone();
+      for (std::set<int>::const_iterator it = indices.begin();
+           it != indices.end(); it++)
+        selection->selectColumn(*it + newBegin, true);
+      indices = selection->getIndices();
+      col     = origCol + newBegin;
+      m_lastCol += newBegin;
+      m_firstCol += newBegin;
+      int currentIndx = app->getCurrentColumn()->getColumnIndex();
+      app->getCurrentColumn()->setColumnIndex(currentIndx + newBegin);
+    } else {
+      int currEnd = xsh->getColumnCount() - 1;
+      int newEnd  = *indices.rbegin() + dCol;
+      if (newEnd > currEnd) {
+        for (int x = currEnd + 1; x <= newEnd; x++) ii.insert(x);
+        ColumnCmd::insertEmptyColumns(ii);
```

Comment:
For now dragging columns or cells to upward (in timeline) will insert empty columns accordingly and there is no way to remove such empty columns other than selecting and deleting them manually.

It is open to discussion, but how about adding one (or some) of the following features?

- Dragging of cells/columns upward can insert only one empty column above for each dragging operation in timeline mode. 

- Each dragging operation will automatically remove the last empty columns on mouse release.

- Add "Remove the Last Empty Columns" command which can remove all the empty columns above the topmost occupied column.

Reply:
I'm leaning towards 2nd or 3rd option.